File: 分支限界算法/max_loading.py
from typing import List


class Node:
    def __init__(self) -> None:
        self.weight: int = 0


_dummy_node: Node = Node()


class MaxLoading:
    def __init__(self, weights: List[int], c1_capacity: int, c2_capacity: int) -> None:
        self._weights: List[int] = weights
        self._c1_capacity: int = c1_capacity
        self._c2_capacity: int = c2_capacity

        # 装完第 i 件物品后，仍然剩余的总重量
        self._remaining: List[int] = [0 for _ in range(len(weights))]
        ind: int = len(self._weights) - 2
        while ind >= 0:
            self._remaining[ind] = self._remaining[ind + 1] + self._weights[ind + 1]
            ind -= 1

    def can_load(self) -> bool:
        if not len(self._weights):
            return True
        # 对第一艘船进行优化，尽量装满它
        queue: List[Node] = [Node(), _dummy_node]
        for ind in range(len(self._weights)):  # type: int
            node: Node = queue.pop(0)
            while node is not _dummy_node:
                # 情况 1：能装下
                if node.weight + self._weights[ind] <= self._c1_capacity:
                    new_node: Node = Node()
                    new_node.weight = node.weight + self._weights[ind]
                    queue.append(new_node)
                # 情况 2：可以不装
                if node.weight + self._weights[ind] + self._remaining[ind] > self._c1_capacity:
                    queue.append(node)
                node = queue.pop(0)
            queue.append(_dummy_node)

        best: int = 0
        for node in queue[:-1]:  # type: Node
            if node.weight > best:
                best = node.weight
        print(best)
        return sum(self._weights) - best <= self._c2_capacity


if __name__ == "__main__":
    print(MaxLoading([20, 28, 25, 25], 70, 30).can_load())

File: 分支限界算法/max_loading.go
package main

import "fmt"

func MaxLoading(weights []int, c1, c2 int) bool {
	if len(weights) == 0 {
		return true
	}
	// remaining[i] 表示装完第 i 个元素之后，仍然剩余的重量
	remaining := make([]int, len(weights))
	sum := weights[len(weights)-1]
	for i := len(weights) - 2; i >= 0; i-- {
		remaining[i] = weights[i+1] + remaining[i+1]
		sum += weights[i]
	}

	// 0 表示重量为 0，-1 表示 Dummy 节点，是每一层的分隔符
	queue := []int{0, -1}
	for ind := 0; ind < len(weights); ind++ {
		node := queue[0]
		queue = queue[1:]
		for node != -1 {
			// 列出所有孩子节点

			// 情况 1：能装下
			if node+weights[ind] <= c1 {
				queue = append(queue, node+weights[ind])
			}
			// 情况 2：可以不装
			if node+weights[ind]+remaining[ind] > c1 {
				queue = append(queue, node)
			}
			node = queue[0]
			queue = queue[1:]
		}
		// 在每一层的末尾防止 Dummy 节点
		queue = append(queue, -1)
	}

	// 获取 c1 最多能装多少
	best := 0
	// 最后一个节点是 Dummy 节点，去掉
	for ind := 0; ind < len(queue)-1; ind++ {
		if queue[ind] > best {
			best = queue[ind]
		}
	}
	fmt.Printf("best: %d\n", best)

	// 总重量减去 c1 必须小于或等于 c2，才能装载
	return sum-best <= c2
}

func main() {
	tests := []struct {
		weights  []int
		c1       int
		c2       int
		expected bool
	}{
		{
			[]int{20, 28, 25, 25},
			70,
			30,
			true,
		},
	}

	for _, test := range tests {
		got := MaxLoading(test.weights, test.c1, test.c2)
		if got != test.expected {
			fmt.Printf("got %v, but %v expected\n", got, test.expected)
			panic("should equal")
		}
	}
}
